runtime.md

```diff
@@ -1,5 +1,47 @@
 # Runtime and Lifecycle
 
+## State
+
+The runtime state for a container is persisted on disk so that external tools can consume and act on this information.
+The runtime state is stored in a JSON encoded file.
+It is recommended that this file is stored in a temporary filesystem so that it can be removed on a system reboot.
+On Linux based systems the state information should be stored in `/run/oci/containers`.
+The directory structure for a container is `/run/oci/containers/<containerID>/state.json`.
+By providing a default location that container state is stored external applications can find all containers running on a system.
+
+* **version** (string) Version of the OCI specification used when creating the container.
+* **id** (string) ID is the container's ID.
+* **pid** (int) Pid is the ID of the main process within the container.
+* **root** (string) Root is the path to the container's bundle directory.
+
+The ID is provided in the state because hooks will be executed with the state as the payload.
+This allows the hook to perform clean and teardown logic after the runtime destroys it's own state.
+The root directory to the bundle is provided in the state so that consumers can find the container's configuration and rootfs.
+
+*Example*
+
+```json
+{
+    "id": "oci-container",
+    "pid": 4422,
+    "root": "/containers/redis"
+}
+```
+
+Linux systems add some platform specific information to the state.
+
+* **externalFds** Paths to the container's open file descriptors.
+
+*Example Linux*
+
+```json
+{
+    "externalFds": [
+        "/proc/33/fd/1"
```

Comment:
On Thu, Aug 27, 2015 at 06:32:33AM -0700, Andrew Vagin wrote:
> It's hard to understand what you are talking about.  I want to show
> an example how it works in libcontainer:

Thanks :).  It helps to have something concrete to make sure we're all
talking about the same things ;).

> "init_process_pid":7725,
> "external_descriptors":["/dev/null","pipe:[66755]","/dev/null"]
>
> # ls -l /proc/7725/fd
> total 0
> lr-x------ 1 root root 64 Aug 27 16:14 0 -> /dev/null
> l-wx------ 1 root root 64 Aug 27 16:14 1 -> pipe:[66755]
> lr-x------ 1 root root 64 Aug 27 16:14 2 -> /dev/null
> l-wx------ 1 root root 64 Aug 27 16:14 255 -> pipe:[66755]

This makes it seem like all the information about open file
descriptors is included in /proc/{pid}/fd.  Are you saying you need
something that's not in /proc?  Or is the issue file descriptor 255,
which you don't have in your external_descriptors list?

> Currently the user creates this pipe and execute "criu restore". But
> CRIU needs to know which file descriptors should be replaced by this
> pipe. For that we use externalFds.

But it seems like you could almost as easily use the values available
via /proc/{pid}/fd.  Why not get them directly from /proc instead of
requiring them to be listed in the state JSON?